README.md
# lolo_webgui
![CI](https://github.com/nilsbore/lolo_webgui/workflows/CI/badge.svg?branch=master) [![license](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

Dashboard interface for the SAM AUV based on roswasm_webgui

## Running the interface

Run using `roslaunch lolo_webgui webgui.launch` and navigate to `1.2.3.4:8080` in a browser (preferrably chrome),
where `1.2.3.4` should be replaced by the IP of the machine that the interface was launched on.

## Dependencies for running (on SAM or elsewhere)

* [roswasm_suite](https://github.com/nilsbore/roswasm_suite): clone in catkin workspace with `git clone https://github.com/nilsbore/roswasm_suite.git`
* [rosbridge_suite](https://github.com/RobotWebTools/rosbridge_suite) after [the commit adding cbor-raw compression](https://github.com/RobotWebTools/rosbridge_suite/commit/dc7fcb282d1326d573abe83579cc7d989ae71739), latest `develop` should do:
  ```bash
  git clone https://github.com/RobotWebTools/rosbridge_suite -b develop # in workspace src folder
  ```
  You can build the workspace as you would otherwise after this, but it won't really do much since the interface is already compiled
  and can be used as-is. You do not need to do any of the following steps if you just want to run the interface.

## Dependencies when developing

* [Emscripten](https://emscripten.org/docs/getting_started/downloads.html) tested with version `1.39.10` but latest should do
* [roswasm_suite](https://github.com/nilsbore/roswasm_suite): clone in catkin workspace with `git clone https://github.com/nilsbore/roswasm_suite.git`
* [rosbridge_suite](https://github.com/RobotWebTools/rosbridge_suite) after [the commit adding cbor-raw compression](https://github.com/RobotWebTools/rosbridge_suite/commit/dc7fcb282d1326d573abe83579cc7d989ae71739), latest `develop` should do:
  ```bash
  git clone https://github.com/RobotWebTools/rosbridge_suite -b develop # in workspace src folder
  ```

## Building when developing

`catkin build` is recommended, since `catkin_make` might leak configurations to other packages.
Make sure to [source the Emscripten environment](https://emscripten.org/docs/getting_started/downloads.html#installation-instructions)
before building the package:
```bash
source /path/to/emsdk/emsdk_env.sh
catkin build
```


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package lolo_webgui_native
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.1.3 (2022-04-05)
------------------
* Bumped package version
* Merge pull request `#2 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/2>`_ from smarc-project-native-gui/merge_upstream
  Merge upstream
* Fixed merge
* Merged upstream
* Updated package.xml
* Update lolo_webgui.rosinstall
* Fixed default native seeting
* Merge remote-tracking branch 'origin/master'
* Native changes
* Merge pull request `#1 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/1>`_ from smarc-project/master
  Fixed bug
* Merge remote-tracking branch 'origin/master'
* Fixed bug
* Update README.md
* New ci configs
* Merge remote-tracking branch 'sam/master' into merge_noetic
* Update lolo_webgui.rosinstall
* Got Lolo webgui working
* Update release.yml
* Replaced sam with lolo everywhere it appears
* Replaced sam with lolo everywhere it appears
* Renamed files
* Replaced sam with lolo everywhere it appears
* Merge pull request `#4 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/4>`_ from nilsbore/master
  Native library
* Merge pull request `#3 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/3>`_ from smarc-project/rename_library
  Rename library
* Renaming library for native
* Merge pull request `#3 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/3>`_ from smarc-project/nilsbore-patch-1
  Update main.yml
* Update main.yml
* Merge pull request `#2 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/2>`_ from smarc-project/2.0
  New roswasm version update
* Merge pull request `#2 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/2>`_ from smarc-project/2.0
  Update to new roswasm version
* Update sam_webgui.rosinstall
* Updated package version
* Fixed some bugs
* Added correct closing of program
* Fixed some bugs
* Updated to new version of roswasm
* Merge pull request `#1 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/1>`_ from nilsbore/master
  Added release building
* Merge pull request `#1 <https://github.com/smarc-project-native-gui/lolo_webgui_native/issues/1>`_ from nilsbore/test_release
  Release building
* Update main.yml
* Update package.xml
* Create release.yml
* Update README.md
* Update README.md
* Update README.md
* Update sam_webgui.rosinstall
* Create main.yml
* Added rules to install precompiled files
* Added precompiled files
* Added install rules
* Got the basics of a CI running for this
* Updated to work with new emscripten
* Completed the sam_webgui package to work with new version of roswasm_suite
* Got rid of a lot of stuff
* Got things to build with catkin tools
* Made having cloned roswasm optional
* Made emscripten optional
* Got it working with separate sam package
* Initial commit
* Contributors: Nils Bore


src/roswasm_lolo.cpp
#include <lolo_webgui/roswasm_lolo.h>

#include <roswasm_webgui/imgui/imgui.h>

namespace roswasm_webgui {

bool draw_thruster_rpm(smarc_msgs::ThrusterRPM& msg, roswasm::Publisher& pub)
{
    ImGui::PushID("RPM cmd slider");
    ImGui::SliderInt("", &msg.rpm, -2000, 2000, "%drpm");
    if (ImGui::IsItemDeactivatedAfterChange()) {
        pub.publish(msg);
    }
    bool lock = ImGui::IsItemActive();

    ImGui::PopID();
    ImGui::SameLine();
    ImGui::InputInt("RPM cmd input", &msg.rpm);
    if (ImGui::IsItemDeactivatedAfterChange()) {
        pub.publish(msg);
    }

    return lock;
}

LoloActuatorWidget::LoloActuatorWidget(roswasm::NodeHandle& nh) : rpm_pub_enabled(false)
{
    //thruster_angles = new TopicPairWidget<geometry_msgs::Pose2D, std_msgs::Float64>(nh, DrawFloatPair("Hori (rad)", -0.1, 0.18, "Vert (rad)", -0.1, 0.15), "core/thrust_vector_cmd", "core/thrust_fb1", "core/thrust_fb2");
    //thruster_rpms = new TopicPairWidget<geometry_msgs::Pose2D, std_msgs::Float64>(nh, DrawFloatPair("Thruster 1", -1000., 1000., "Thruster 2", -1000., 1000.), "core/rpm_cmd", "core/rpm_fb1", "core/rpm_fb2");
    //thruster_angles = new TopicWidget<lolo_msgs::ThrusterAngles>(nh, &draw_thruster_angles, "core/thrust_vector_cmd");
    rudder_angles = new TopicWidget<std_msgs::Float32>(nh, DrawFloat32(-.6, .6), "core/rudder_cmd");
    elevator_angle = new TopicWidget<std_msgs::Float32>(nh, DrawFloat32(-.6, .6), "core/elevator_cmd");
    elevon_port_angle = new TopicWidget<std_msgs::Float32>(nh, DrawFloat32(-.6, .6), "core/elevon_port_cmd");
    elevon_stbd_angle = new TopicWidget<std_msgs::Float32>(nh, DrawFloat32(-.6, .6), "core/elevon_strb_cmd");

    vbs_front_port = new TopicWidget<std_msgs::Float32>(nh, DrawFloat32(0., 100.), "core/vbs_front_port_cmd", "core/vbs_front_port_fb");
    vbs_front_stbd = new TopicWidget<std_msgs::Float32>(nh, DrawFloat32(0., 100.), "core/vbs_front_stbd_cmd", "core/vbs_front_stbd_fb");
    vbs_back_port = new TopicWidget<std_msgs::Float32>(nh, DrawFloat32(0., 100.), "core/vbs_back_port_cmd", "core/vbs_back_port_fb");
    vbs_back_stbd = new TopicWidget<std_msgs::Float32>(nh, DrawFloat32(0., 100.), "core/vbs_back_stbd_cmd", "core/vbs_back_stbd_fb");

    //thruster_rpms = new TopicWidget<lolo_msgs::ThrusterRPMs>(nh, &draw_thruster_rpms, "core/rpm_cmd", "core/rpm_fb");
    thruster1_rpm = new TopicWidget<smarc_msgs::ThrusterRPM>(nh, &draw_thruster_rpm, "core/thruster1_cmd");
    thruster2_rpm = new TopicWidget<smarc_msgs::ThrusterRPM>(nh, &draw_thruster_rpm, "core/thruster2_cmd");
    rpm1_pub = nh.advertise<smarc_msgs::ThrusterRPM>("core/thruster1_cmd", 1000);
    rpm2_pub = nh.advertise<smarc_msgs::ThrusterRPM>("core/thruster2_cmd", 1000);

    pub_timer = nh.createTimer(roswasm::Duration(0.08), std::bind(&LoloActuatorWidget::pub_callback, this, std::placeholders::_1));
    pub_timer.stop();
}

void LoloActuatorWidget::pub_callback(const ros::TimerEvent& e)
{
    if (rpm_pub_enabled) {
        rpm1_pub.publish(thruster1_rpm->get_msg());
        rpm2_pub.publish(thruster2_rpm->get_msg());
    }
}

void LoloActuatorWidget::show_window(bool& show_actuator_window)
{
    ImGui::Begin("Actuator controls", &show_actuator_window);

    if (ImGui::CollapsingHeader("Rudder Angles", ImGuiTreeNodeFlags_Framed | ImGuiTreeNodeFlags_DefaultOpen)) {
        ImGui::PushID("RPMs");
        ImGui::PushID("First");
        ImGui::Text("Rudders");
        ImGui::SameLine();
        rudder_angles->show_widget();
        ImGui::PopID();
        ImGui::PushID("Second");
        ImGui::Text("Elevator");
        ImGui::SameLine();
        elevator_angle->show_widget();
        ImGui::PopID();
        ImGui::PushID("Third");
        ImGui::Text("Port Elevon");
        ImGui::SameLine();
        elevon_port_angle->show_widget();
        ImGui::PopID();
        ImGui::PushID("Fourth");
        ImGui::Text("Starboard Elevon");
        ImGui::SameLine();
        elevon_stbd_angle->show_widget();
        ImGui::PopID();
        ImGui::PopID();
    }
    if (ImGui::CollapsingHeader("Thruster RPMs", ImGuiTreeNodeFlags_Framed | ImGuiTreeNodeFlags_DefaultOpen)) {
        ImGui::PushID("RPMs");
        ImGui::PushID("First");
        ImGui::Text("Thruster 1");
        ImGui::SameLine();
        thruster1_rpm->show_widget();
        ImGui::PopID();
        ImGui::PushID("Second");
        ImGui::Text("Thruster 2");
        ImGui::SameLine();
        thruster2_rpm->show_widget();
        ImGui::PopID();
        ImGui::Checkbox("Publish RPMs at 10hz", &rpm_pub_enabled);
        ImGui::PopID();
        if (rpm_pub_enabled) { // && pub_timer == nullptr) {
            pub_timer.start();
        }
        else { //if (!rpm_pub_enabled && pub_timer != nullptr) {
            pub_timer.stop();
        }
    }
    if (ImGui::CollapsingHeader("VBS", ImGuiTreeNodeFlags_Framed | ImGuiTreeNodeFlags_DefaultOpen)) {
        ImGui::PushID("VBS");
        ImGui::PushID("First");
        ImGui::Text("Front port");
        ImGui::SameLine();
        vbs_front_port->show_widget();
        ImGui::PopID();
        ImGui::PushID("Second");
        ImGui::Text("Front starboard");
        ImGui::SameLine();
        vbs_front_stbd->show_widget();
        ImGui::PopID();
        ImGui::PushID("Third");
        ImGui::Text("Back port");
        ImGui::SameLine();
        vbs_back_port->show_widget();
        ImGui::PopID();
        ImGui::PushID("Fourth");
        ImGui::Text("Back starboard");
        ImGui::SameLine();
        vbs_back_stbd->show_widget();
        ImGui::PopID();
        ImGui::PopID();
    }

    ImGui::End();
}

LoloDashboardWidget::LoloDashboardWidget(roswasm::NodeHandle& nh) : was_leak(false)
{
    leak = new TopicBuffer<smarc_msgs::Leak>(nh, "core/leak_fb");
    gps = new TopicBuffer<sensor_msgs::NavSatFix>(nh, "core/gps");
    battery = new TopicBuffer<sensor_msgs::BatteryState>(nh, "core/battery");
    odom = new TopicBuffer<nav_msgs::Odometry>(nh, "dr/odom", 1000);
    //rpms = new TopicBuffer<lolo_msgs::ThrusterRPMs>(nh, "core/rpm_fb", 1000);
    rpm1 = new TopicBuffer<smarc_msgs::ThrusterRPM>(nh, "core/thruster1_cmd", 1000);
    rpm2 = new TopicBuffer<smarc_msgs::ThrusterRPM>(nh, "core/thruster2_cmd", 1000);
    depth = new TopicBuffer<std_msgs::Float64>(nh, "ctrl/depth_feedback", 1000);
    pitch = new TopicBuffer<std_msgs::Float64>(nh, "ctrl/pitch_feedback", 1000);
    roll = new TopicBuffer<std_msgs::Float64>(nh, "ctrl/roll_feedback", 1000);
    yaw = new TopicBuffer<std_msgs::Float64>(nh, "ctrl/yaw_feedback", 1000);
}

void LoloDashboardWidget::show_window(bool& show_dashboard_window)
{
    ImGui::Begin("Status dashboard", &show_dashboard_window);

    if (ImGui::CollapsingHeader("Critical Info", ImGuiTreeNodeFlags_Framed | ImGuiTreeNodeFlags_DefaultOpen)) {
        was_leak = was_leak || leak->get_msg().value;

        float sz = ImGui::GetTextLineHeight();
        std::string status_text;
        ImColor status_color;
        if (!was_leak) {
            status_text = "No leaks!";
            status_color = ImColor(0, 255, 0);
        }
        else {
            status_text = "Leak!!!!!";
            status_color = ImColor(255, 0, 0);
        }
        ImVec2 p = ImGui::GetCursorScreenPos();
        ImGui::GetWindowDrawList()->AddRectFilled(p, ImVec2(p.x+sz, p.y+sz), status_color);
        ImGui::Dummy(ImVec2(sz, sz));
        ImGui::SameLine();
        ImGui::Text("%s", status_text.c_str());

        ImGui::SameLine(150);
        ImGui::Text("Battery: %.0f%%", battery->get_msg().percentage);
    }

    if (ImGui::CollapsingHeader("GPS and depth", ImGuiTreeNodeFlags_Framed | ImGuiTreeNodeFlags_DefaultOpen)) {
        ImGui::Text("Lat: %.5f", gps->get_msg().latitude);
        ImGui::SameLine(150);
        ImGui::Text("Lon: %.5f", gps->get_msg().longitude);
        ImGui::SameLine(300);
        ImGui::Text("Depth: %.2fm", depth->get_msg().data);
    }

    if (ImGui::CollapsingHeader("DR translation", ImGuiTreeNodeFlags_Framed | ImGuiTreeNodeFlags_DefaultOpen)) {
        ImGui::Text("X: %.2fm", odom->get_msg().pose.pose.position.x);
        ImGui::SameLine(150);
        ImGui::Text("Y: %.2fm", odom->get_msg().pose.pose.position.y);
        ImGui::SameLine(300);
        ImGui::Text("Z: %.2fm", odom->get_msg().pose.pose.position.z);
    }

    if (ImGui::CollapsingHeader("DR rotation", ImGuiTreeNodeFlags_Framed | ImGuiTreeNodeFlags_DefaultOpen)) {
        ImGui::Text("Roll: %.2fdeg", 180./M_PI*roll->get_msg().data);
        ImGui::SameLine(150);
        ImGui::Text("Pitch: %.2fdeg", 180./M_PI*pitch->get_msg().data);
        ImGui::SameLine(300);
        ImGui::Text("Yaw: %.2fdeg", 180./M_PI*yaw->get_msg().data);
    }

    if (ImGui::CollapsingHeader("Actuator feedback", ImGuiTreeNodeFlags_Framed | ImGuiTreeNodeFlags_DefaultOpen)) {
        ImGui::Text("RPMs: %d, %d rpm", rpm1->get_msg().rpm, rpm2->get_msg().rpm);
    }

    ImGui::End();
}

LoloTeleopWidget::LoloTeleopWidget(roswasm::NodeHandle& nh) : enabled(false) //, pub_timer(nullptr)
{
    rudder_pub = nh.advertise<std_msgs::Float32>("core/rudder_cmd", 1000);
    elevator_pub = nh.advertise<std_msgs::Float32>("core/elevator_cmd", 1000);
    elevon_port_pub = nh.advertise<std_msgs::Float32>("core/elevon_port_cmd", 1000);
    elevon_stbd_pub = nh.advertise<std_msgs::Float32>("core/elevon_strb_cmd", 1000);
    rpm1_pub = nh.advertise<smarc_msgs::ThrusterRPM>("core/thruster1_cmd", 1000);
    rpm2_pub = nh.advertise<smarc_msgs::ThrusterRPM>("core/thruster2_cmd", 1000);
    pub_timer = nh.createTimer(roswasm::Duration(0.08), std::bind(&LoloTeleopWidget::pub_callback, this, std::placeholders::_1));
    pub_timer.stop();
}

void LoloTeleopWidget::pub_callback(const ros::TimerEvent& e)
{
    if (enabled) {
        //angle_pub.publish(angles_msg);
        elevon_port_pub.publish(elevon_msg);
        elevon_msg.data *= -1.;
        elevon_stbd_pub.publish(elevon_msg);
        elevator_pub.publish(elevon_msg);
        elevon_msg.data *= -1.;
        rpm1_pub.publish(rpm1_msg);
        rpm2_pub.publish(rpm2_msg);
        rudder_pub.publish(rudder_msg);
    }
}

void LoloTeleopWidget::show_window(bool& show_teleop_window)
{
    ImGuiIO& io = ImGui::GetIO();
    ImVec4 col = ImGui::GetStyle().Colors[23];

    ImGui::SetNextWindowSize(ImVec2(472, 80), ImGuiCond_FirstUseEver);
    ImGui::Begin("Keyboard teleop", &show_teleop_window);

    rudder_msg.data = elevon_msg.data = rpm1_msg.rpm = rpm2_msg.rpm = 0.0f;

    float sz = ImGui::GetTextLineHeight();
    ImGui::BeginGroup();
    ImGui::BeginGroup();
    ImGui::Dummy(ImVec2(sz, 0.75f*sz));
    ImGui::Checkbox("Teleop enabled", &enabled);
    if (enabled) {
        pub_timer.start();
    }
    else { //if (!enabled && pub_timer != nullptr) {
        pub_timer.stop();
    }
    ImGui::EndGroup();

    ImGui::SameLine();

    ImGui::BeginGroup();
    ImGui::Dummy(ImVec2(sz, 1.5f*sz));
    bool key_down = enabled && io.KeysDownDuration[263] >= 0.0f;
    if (key_down) {
        ImGui::PushStyleColor(ImGuiCol_Button, col);
    }
    if (ImGui::ArrowButton("##left", ImGuiDir_Left) || key_down) {
        //angles_msg.thruster_horizontal_radians = -0.10;
        rudder_msg.data = -0.6;
    }
    if (key_down) {
        ImGui::PopStyleColor();   
    }
    ImGui::EndGroup();
    ImGui::SameLine();
    ImGui::BeginGroup();
    key_down = enabled && io.KeysDownDuration[265] >= 0.0f;
    if (key_down) {
        ImGui::PushStyleColor(ImGuiCol_Button, col);
    }
    if (ImGui::ArrowButton("##up", ImGuiDir_Up) || key_down) {
        //angles_msg.thruster_vertical_radians = 0.10;
        elevon_msg.data = -0.6;
    }
    if (key_down) {
        ImGui::PopStyleColor();   
    }
    key_down = enabled && io.KeysDownDuration[264] >= 0.0f;
    if (key_down) {
        ImGui::PushStyleColor(ImGuiCol_Button, col);
    }
    if (ImGui::ArrowButton("##down", ImGuiDir_Down) || key_down) {
        //angles_msg.thruster_vertical_radians = -0.10;
        elevon_msg.data = 0.6;
    }
    if (key_down) {
        ImGui::PopStyleColor();   
    }
    ImGui::EndGroup();
    ImGui::SameLine();
    ImGui::BeginGroup();
    ImGui::Dummy(ImVec2(sz, 1.5f*sz));
    key_down = enabled && io.KeysDownDuration[262] >= 0.0f;
    if (key_down) {
        ImGui::PushStyleColor(ImGuiCol_Button, col);
    }
    if (ImGui::ArrowButton("##right", ImGuiDir_Right) || key_down) {
        //angles_msg.thruster_horizontal_radians = 0.10;
        rudder_msg.data = 0.6;
    }
    if (key_down) {
        ImGui::PopStyleColor();   
    }
    ImGui::EndGroup();

    ImGui::SameLine();

    ImGui::BeginGroup();
    ImGui::Dummy(ImVec2(sz, sz));
    ImGui::Text("Thrust Vector");
    ImGui::EndGroup();

    ImGui::SameLine();

    ImGui::BeginGroup();
    key_down = enabled && io.KeysDownDuration[87] >= 0.0f;
    if (key_down) {
        ImGui::PushStyleColor(ImGuiCol_Button, col);
    }
    if (ImGui::Button("w") || key_down) {
        rpm1_msg.rpm = rpm2_msg.rpm = 1000;
    }
    if (key_down) {
        ImGui::PopStyleColor();   
    }
    key_down = enabled && io.KeysDownDuration[83] >= 0.0f;
    if (key_down) {
        ImGui::PushStyleColor(ImGuiCol_Button, col);
    }
    if (ImGui::Button("s") || key_down) {
        rpm1_msg.rpm = rpm2_msg.rpm = -1000;
    }
    if (key_down) {
        ImGui::PopStyleColor();   
    }
    ImGui::EndGroup();
    ImGui::SameLine();
    ImGui::BeginGroup();
    ImGui::Text("Forward");
    ImGui::Dummy(ImVec2(sz, 0.5f*sz));
    ImGui::Text("Reverse");
    ImGui::EndGroup();
    ImGui::EndGroup();

    ImGui::End();

    //ImGui::Text("Keys down:");      for (int i = 0; i < IM_ARRAYSIZE(io.KeysDown); i++) if (io.KeysDownDuration[i] >= 0.0f)     { ImGui::SameLine(); ImGui::Text("%d (%.02f secs)", i, io.KeysDownDuration[i]); }
}


} // namespace roswasm_webgui


src/main.cpp
#include <stdio.h>

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#endif

#include <roswasm_webgui/imgui/imgui.h>
#include <roswasm_webgui/imgui/imgui_impl_glfw.h>
#include <roswasm_webgui/imgui/imgui_impl_opengl3.h>

#ifdef ROSWASM_NATIVE
#include <GL/glew.h> 
#else
#define GLFW_INCLUDE_ES3
#include <GLES3/gl3.h>
#endif
#include <GLFW/glfw3.h>

#include <roswasm/roswasm.h>
#include <roswasm_webgui/roswasm_monlaunch.h>
#include <roswasm_webgui/roswasm_image.h>
#include <roswasm_webgui/roswasm_examples.h>

#include <lolo_webgui/roswasm_lolo.h>

roswasm::NodeHandle* nh; 
roswasm_webgui::MonlaunchWidget* monlaunch_widget;
roswasm_webgui::ImageWidget* image_widget;
roswasm_webgui::LoloActuatorWidget* actuator_widget;
roswasm_webgui::LoloDashboardWidget* dashboard_widget;
roswasm_webgui::LoloTeleopWidget* teleop_widget;

GLFWwindow* g_window;
ImVec4 clear_color = ImVec4(0.25f, 0.45f, 0.55f, 1.00f);
ImVec4 emergency_color = ImVec4(1.0f, 0.0f, 0.0f, 1.00f);
ImGuiContext* imgui = 0;
bool show_demo_window = false;
bool show_monlaunch_window = true;
bool show_image_window = true;
bool show_actuator_window = true;
bool show_dashboard_window = true;
bool show_teleop_window = true;

#ifndef ROSWASM_NATIVE
EM_JS(int, canvas_get_width, (), {
    return Module.canvas.width;
});

EM_JS(int, canvas_get_height, (), {
    return Module.canvas.height;
});

EM_JS(void, resizeCanvas, (), {
    js_resizeCanvas();
});
#endif

void loop()
{
    if (glfwWindowShouldClose(g_window)) {
        roswasm::shutdown();
        return;
    }

#ifndef ROSWASM_NATIVE
    glfwSetWindowSize(g_window, canvas_get_width(), canvas_get_height());
#endif
    glfwPollEvents();

    //ImGui::SetCurrentContext(imgui);

    ImGui_ImplOpenGL3_NewFrame();
    ImGui_ImplGlfw_NewFrame();
    ImGui::NewFrame();

    {
        ImGui::SetNextWindowPos(ImVec2(30, 30), ImGuiCond_FirstUseEver);
        ImGui::SetNextWindowSize(ImVec2(482, 210), ImGuiCond_FirstUseEver);
        ImGui::Begin("Roswasm webgui"); //, &show_another_window);
        float sz = ImGui::GetTextLineHeight();
        std::string status_text;
        ImColor status_color;
        if (nh->ok()) {
#ifdef ROSWASM_NATIVE
            status_text = "ROS OK";
#else
            status_text = "Connected to " + nh->get_websocket_url();
#endif
            status_color = ImColor(0, 255, 0);
        }
        else {
#ifdef ROSWASM_NATIVE
            status_text = "ROS NOT OK!";
#else
            status_text = "Disconnected from " + nh->get_websocket_url();
#endif
            status_color = ImColor(255, 0, 0);
        }
        ImVec2 p = ImGui::GetCursorScreenPos();
        ImGui::GetWindowDrawList()->AddRectFilled(p, ImVec2(p.x+sz, p.y+sz), status_color);
        ImGui::Dummy(ImVec2(sz, sz));
        ImGui::SameLine();
        ImGui::Text("%s", status_text.c_str());

        ImGui::Checkbox("Launch control", &show_monlaunch_window);
        ImGui::Checkbox("Image topic", &show_image_window);
        ImGui::Checkbox("Actuator controls", &show_actuator_window);
        ImGui::Checkbox("Status dashboard", &show_dashboard_window);
        ImGui::Checkbox("Keyboard teleop", &show_teleop_window);
        ImGui::Checkbox("Demo widgets", &show_demo_window);      // Edit bools storing our windows open/close state

        //ImGui::Text("Application average %.3f ms/frame (%.1f FPS)", 1000.0f / ImGui::GetIO().Framerate, ImGui::GetIO().Framerate);
        ImGui::ColorEdit3("Background", (float*)&clear_color); // Edit 3 floats representing a color
        ImGui::End();
    }

    if (show_monlaunch_window) {
        ImGui::SetNextWindowPos(ImVec2(30, 270), ImGuiCond_FirstUseEver);
        ImGui::SetNextWindowSize(ImVec2(482, 511), ImGuiCond_FirstUseEver);
        monlaunch_widget->show_window(show_monlaunch_window);
    }

    if (show_image_window) {
        ImGui::SetNextWindowPos(ImVec2(1072, 30), ImGuiCond_FirstUseEver);
        ImGui::SetNextWindowSize(ImVec2(472, 427), ImGuiCond_FirstUseEver);
        image_widget->show_window(show_image_window);
    }

    if (show_demo_window) {
        ImGui::SetNextWindowPos(ImVec2(600, 60), ImGuiCond_FirstUseEver);
        ImGui::ShowDemoWindow(&show_demo_window);
    }

    if (show_actuator_window) {
        ImGui::SetNextWindowPos(ImVec2(542, 303), ImGuiCond_FirstUseEver);
        ImGui::SetNextWindowSize(ImVec2(500, 478), ImGuiCond_FirstUseEver);
        actuator_widget->show_window(show_actuator_window);
    }

    if (show_dashboard_window) {
        ImGui::SetNextWindowPos(ImVec2(542, 30), ImGuiCond_FirstUseEver);
        ImGui::SetNextWindowSize(ImVec2(500, 243), ImGuiCond_FirstUseEver);
        dashboard_widget->show_window(show_dashboard_window);
    }

    if (show_teleop_window) {
        ImGui::SetNextWindowPos(ImVec2(1072, 487), ImGuiCond_FirstUseEver);
        teleop_widget->show_window(show_teleop_window);
    }

    ImGui::Render();

    int display_w, display_h;
    //glfwMakeContextCurrent(g_window);
    glfwGetFramebufferSize(g_window, &display_w, &display_h);
    glViewport(0, 0, display_w, display_h);
    if (dashboard_widget->is_emergency()) {
        glClearColor(emergency_color.x, emergency_color.y, emergency_color.z, emergency_color.w);
    }
    else {
        glClearColor(clear_color.x, clear_color.y, clear_color.z, clear_color.w);
    }
    glClear(GL_COLOR_BUFFER_BIT);

    ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
    //glfwMakeContextCurrent(g_window);
    glfwSwapBuffers(g_window);
}

static void glfw_error_callback(int error, const char* description)
{
    fprintf(stderr, "Glfw Error %d: %s\n", error, description);
}

int init()
{
    glfwSetErrorCallback(glfw_error_callback);
    if( !glfwInit() )
    {
        fprintf( stderr, "Failed to initialize GLFW\n" );
        return 1;
    }

    //glfwWindowHint(GLFW_LOLOPLES, 4); // 4x antialiasing
    //glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE); // We don't want the old OpenGL
    // GL 3.0 + GLSL 130
    const char* glsl_version = "#version 130";
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);

    // Open a window and create its OpenGL context
    int canvasWidth = 1280;
    int canvasHeight = 720;
    g_window = glfwCreateWindow(canvasWidth, canvasHeight, "ROSWASM Example", NULL, NULL);
    if( g_window == NULL )
    {
        fprintf( stderr, "Failed to open GLFW window.\n" );
        glfwTerminate();
        return -1;
    }

    glfwMakeContextCurrent(g_window); // Initialize GLEW


#ifdef ROSWASM_NATIVE
    glfwSwapInterval(1); // Enable vsync

    // Initialize OpenGL loader
    bool err = glewInit() != GLEW_OK;
    if (err)
    {
        fprintf(stderr, "Failed to initialize OpenGL loader!\n");
        return 1;
    }
#endif

    // Create game objects
    // Setup Dear ImGui binding
    IMGUI_CHECKVERSION();
    ImGui::CreateContext();
    ImGuiIO& io = ImGui::GetIO(); (void)io;

    // Setup style
    ImGui::StyleColorsLight();
    //ImGui::StyleColorsDark();
    //ImGui::StyleColorsClassic();

    //ImGui_ImplGlfw_InitForOpenGL(g_window, false);
    ImGui_ImplGlfw_InitForOpenGL(g_window, true);
    //ImGui_ImplOpenGL3_Init();
    ImGui_ImplOpenGL3_Init(glsl_version);

    // Load Fonts
    io.Fonts->AddFontDefault();
    /*
    io.Fonts->AddFontFromFileTTF("data/xkcd-script.ttf", 23.0f);
    io.Fonts->AddFontFromFileTTF("data/xkcd-script.ttf", 18.0f);
    io.Fonts->AddFontFromFileTTF("data/xkcd-script.ttf", 26.0f);
    io.Fonts->AddFontFromFileTTF("data/xkcd-script.ttf", 32.0f);
    */

    //imgui =  ImGui::GetCurrentContext();

    // Cursor callbacks
    /*
    glfwSetMouseButtonCallback(g_window, ImGui_ImplGlfw_MouseButtonCallback);
    glfwSetScrollCallback(g_window, ImGui_ImplGlfw_ScrollCallback);
    glfwSetKeyCallback(g_window, ImGui_ImplGlfw_KeyCallback);
    glfwSetCharCallback(g_window, ImGui_ImplGlfw_CharCallback);
    */

#ifndef ROSWASM_NATIVE
    resizeCanvas();
#endif

    return 0;
}

void quit()
{
    // Cleanup
    ImGui_ImplOpenGL3_Shutdown();
    ImGui_ImplGlfw_Shutdown();
    ImGui::DestroyContext();

    glfwDestroyWindow(g_window);
    glfwTerminate();
}

extern "C" int main(int argc, char** argv)
{
    if (init() != 0) return 1;

    roswasm::init(argc, argv, "roswasm_webgui");

#ifdef ROSWASM_NATIVE
    nh = new roswasm::NodeHandle();
#else
    if (argc < 3) {
        printf("Rosbridge server ip and port not provided!");
        return 1;
    }

    std::string rosbridge_ip(argv[1]);
    std::string rosbridge_port(argv[2]);
    nh = new roswasm::NodeHandle(rosbridge_ip, rosbridge_port);
#endif

    monlaunch_widget = new roswasm_webgui::MonlaunchWidget(*nh);
    image_widget = new roswasm_webgui::ImageWidget(*nh);
    actuator_widget = new roswasm_webgui::LoloActuatorWidget(*nh);
    dashboard_widget = new roswasm_webgui::LoloDashboardWidget(*nh);
    teleop_widget = new roswasm_webgui::LoloTeleopWidget(*nh);

    roswasm::Duration loop_rate(1./20.);
    roswasm::spinLoop(loop, loop_rate);
    //roswasm::spinLoop(loop);

    quit();

    return 0;
}
